internal/controllers/reconciliation/controller.go

```diff
@@ -266,81 +261,48 @@ func (c *Controller) reconcileResource(ctx context.Context, comp *apiv1.Composit
+		patch, err := json.Marshal(&resource.Patch)
 		if err != nil {
-			return false, fmt.Errorf("adding resource version: %w", err)
+			return false, fmt.Errorf("encoding json patch: %w", err)
 		}
-	}
-	if len(patch) == 0 {
-		logger.V(1).Info("skipping empty patch")
-		return false, nil
-	}
-	reconciliationActions.WithLabelValues("patch").Inc()
-	if insecureLogPatch {
-		logger.V(1).Info("INSECURE logging patch", "patch", string(patch))
-	}
-	err = c.upstreamClient.Patch(ctx, current, client.RawPatch(patchType, patch))
-	if err != nil {
-		return false, fmt.Errorf("applying patch: %w", err)
-	}
-	logger.V(0).Info("patched resource", "patchType", string(patchType), "resourceVersion", current.GetResourceVersion(), "previousResourceVersion", prevRV)
-
-	return true, nil
-}
 
-func (c *Controller) buildPatch(ctx context.Context, prev, next *reconstitution.Resource, current *unstructured.Unstructured) ([]byte, types.PatchType, error) {
-	if next.Patch != nil {
-		if !next.NeedsToBePatched(current) {
-			return []byte{}, types.JSONPatchType, nil
+		err = c.upstreamClient.Patch(ctx, current, client.RawPatch(types.JSONPatchType, patch))
+		if err != nil {
+			return false, fmt.Errorf("applying patch: %w", err)
 		}
-		patch, err := json.Marshal(&next.Patch)
-		return patch, types.JSONPatchType, err
-	}
 
-	prevJS, err := prev.Finalize()
-	if err != nil {
-		return nil, "", reconcile.TerminalError(fmt.Errorf("building json representation of previous state: %w", err))
-	}
-
-	nextJS, err := next.Finalize()
-	if err != nil {
-		return nil, "", reconcile.TerminalError(fmt.Errorf("building json representation of next state: %w", err))
+		reconciliationActions.WithLabelValues("patch").Inc()
+		logger.V(0).Info("patched resource", "resourceVersion", current.GetResourceVersion())
+		return true, nil
 	}
 
-	currentJS, err := current.MarshalJSON()
+	// Compute a merge patch
+	updated, typed, err := resource.Merge(ctx, prev, current, c.discovery)
 	if err != nil {
-		return nil, "", reconcile.TerminalError(fmt.Errorf("building json representation of current state: %w", err))
+		return false, fmt.Errorf("performing three-way merge: %w", err)
 	}
-
-	model, err := c.discovery.Get(ctx, next.GVK)
-	if err != nil {
-		return nil, "", fmt.Errorf("getting merge metadata: %w", err)
+	if updated == nil {
+		logger.V(1).Info("skipping empty update")
+		return false, nil
 	}
-
-	// FIXME: This is a very nasty hack which should not be needed once we have
-	// support for semantic equality checks.
-	pdbGVK := schema.GroupVersionKind{Group: "policy", Version: "v1", Kind: "PodDisruptionBudget"}
-	if model == nil || (next != nil && next.GVK == pdbGVK) {
-		patch, err := jsonmergepatch.CreateThreeWayJSONMergePatch(prevJS, nextJS, currentJS)
-		if err != nil {
-			return nil, "", reconcile.TerminalError(err)
-		}
-		return patch, types.MergePatchType, err
+	if insecureLogPatch {
+		js, _ := updated.MarshalJSON()
+		logger.V(1).Info("INSECURE logging patch", "update", string(js))
```

Comment:
I'll follow up with a PR to improve the logging here. Right now it just logs the entire resource.